cajita/src/Cajita_ParticleGridDistributor.hpp

```diff
@@ -0,0 +1,390 @@
+        {
+            local_low[d] = local_mesh.lowCorner( Cajita::Own(), d );
+            local_high[d] = local_mesh.highCorner( Cajita::Own(), d );
+            periodic[d] = global_grid.isPeriodic( d );
+            global_low[d] = global_mesh.lowCorner( d );
+            global_high[d] = global_mesh.highCorner( d );
+            global_extent[d] = global_mesh.extent( d );
+        }
+
+        Kokkos::parallel_for(
+            "get_migrate_destinations",
+            Kokkos::RangePolicy<execution_space>( 0, positions.size() ),
+            KOKKOS_LAMBDA( const int p ) {
+                // Compute the logical index of the neighbor we are sending to.
+                int nid[num_space_dim];
+                for ( std::size_t d = 0; d < num_space_dim; ++d )
+                {
+                    nid[d] = 1;
+                    if ( positions( p, d ) < local_low[d] )
+                        nid[d] = 0;
+                    else if ( positions( p, d ) > local_high[d] )
+                        nid[d] = 2;
+                }
+
+                // Compute the destination MPI rank [ni + 3*(nj + 3*nk) in 3d].
+                int neighbor_index = nid[0];
+                for ( std::size_t d = 1; d < num_space_dim; ++d )
+                {
+                    int npower = 1;
+                    for ( std::size_t dp = 1; dp <= d; ++dp )
+                        npower *= 3;
+                    neighbor_index += npower * nid[d];
+                }
+                destinations( p ) = neighbor_ranks( neighbor_index );
+
+                // Shift particles through periodic boundaries.
+                for ( std::size_t d = 0; d < num_space_dim; ++d )
+                {
+                    if ( periodic[d] )
+                    {
+                        if ( positions( p, d ) > global_high[d] )
+                            positions( p, d ) -= global_extent[d];
+                        else if ( positions( p, d ) < global_low[d] )
+                            positions( p, d ) += global_extent[d];
+                    }
+                }
+            } );
+    }
+
+  private:
+    LocalGridType _local_grid;
+};
+
+//-----------------------------------------------------------------------//
+/*!
+  \brief Check for the number of particles that must be communicated
+
+  \tparam LocalGridType Cajita LocalGrid type.
+
+  \tparam PositionSliceType Particle position type.
+
+  \param local_grid The local grid containing periodicity and system bound
+  information.
+
+  \param positions The particle position container, either Slice or View.
+
+  \param minimum_halo_width Number of halo mesh widths to include for
+  ghosting.
+*/
+template <class LocalGridType, class PositionSliceType>
+int migrateCount( const LocalGridType& local_grid,
+                  const PositionSliceType& positions,
+                  const int minimum_halo_width )
+{
+    ParticleGridDistributor<LocalGridType> pg_distributor( local_grid );
```

Comment:
I don't think `ParticleGridDistributor` needs to be an object - the code in `migrateCount` can just be added here instead as a free function. You should have some kind of `ParticleGridDistributor` namespace though for these free functions.

Reply:
Or instead of a namespace prefix the functions with `particleGridDistributor` as you have below.